Menus/ScenarioGUI.cpp

```diff
@@ -26,2022 +12,1051 @@
+				m_PrevMousePos = mousePos;
+
+				KeepBoxInScreenBounds(m_DraggedBox);
+				if (m_DraggedBox == m_ScenarioCollectionBoxes.at(ScenarioCollections::SceneInfoBox)) { CalculateLinesToSitePoint(); }
+			} else {
+				m_DraggedBox = nullptr;
+			}
+
+			if (g_UInputMan.MenuButtonPressed(UInputMan::MENU_EITHER)) {
+				if (m_HoveredScene) {
+					SetSelectedScene(m_HoveredScene);
+					g_GUISound.ItemChangeSound()->Play();
+				}
+
+				GUICollectionBox *hoveredBox = dynamic_cast<GUICollectionBox *>(m_ScenarioGUIController->GetControlUnderPoint(mouseX, mouseY, m_ScenarioCollectionBoxes.at(ScenarioCollections::RootBox), 1));
+				const GUIControl *hoveredControl = m_ScenarioGUIController->GetControlUnderPoint(mouseX, mouseY, hoveredBox, 1);
+				bool nonDragControl = (dynamic_cast<const GUIButton *>(hoveredControl) || dynamic_cast<const GUISlider *>(hoveredControl) || dynamic_cast<const GUIComboBox *>(hoveredControl));
+				if (hoveredBox && !nonDragControl && !m_DraggedBox && !m_ActivitySelectComboBox->IsDropped()) {
+					m_DraggedBox = hoveredBox;
+					m_PrevMousePos = mousePos;
+				}
+			}
+		}
+
+		m_ScenarioGUIController->Update();
+
+		///////////////////////////////////////
+		// Handle events
+
+		GUIEvent guiEvent;
+		while (m_ScenarioGUIController->GetEvent(&guiEvent)) {
+			const GUIControl *eventControl = guiEvent.GetControl();
+			std::string eventControlName = guiEvent.GetControl()->GetName();
+
+			if (m_ScenarioCollectionBoxes.at(ScenarioCollections::ActivitySelectBox)->GetVisible()) {
+				if (guiEvent.GetType() == GUIEvent::Command) {
+					if (eventControlName == "BackToMainButton") {
+						HideAllScreens();
+						g_GUISound.BackButtonPressSound()->Play();
+						return ScenarioUpdateResult::BackToMain;
+					} else if (eventControlName == "ButtonResume") {
+						g_GUISound.BackButtonPressSound()->Play();
+						return ScenarioUpdateResult::ActivityResumed;
+					}
+
+					if (m_ScenarioCollectionBoxes.at(ScenarioCollections::SceneInfoBox)->GetVisible()) {
+						if (eventControl == m_ScenarioButtons.at(ScenarioButtons::StartHereButton)) {
+							HideAllScreens();
+							ShowPlayersBox();
+							g_GUISound.ButtonPressSound()->Play();
+						} else if (eventControl == m_SceneCloseButton) {
+							SetSelectedScene(nullptr);
+							g_GUISound.ButtonPressSound()->Play();
+						}
+					}
+				} else if (guiEvent.GetType() == GUIEvent::Notification) {
+					if (eventControl == m_ActivitySelectComboBox && guiEvent.GetMsg() == GUIComboBox::Closed) {
+						const Activity *activity = dynamic_cast<const Activity *>(m_ActivitySelectComboBox->GetSelectedItem()->m_pEntity);
+						if (activity != m_SelectedActivity) {
+							m_SelectedActivity = activity;
+							// The activity selection has changed.
+							UpdateActivityBox();
+
+							// Deselect any previously selected scene. it may not be compatible with the new activity.
```

Comment:
I'm on the fence about this comment, the code is not actually complicated, but it's slightly ugly, and this does explain why it's called, which is good. If you wanna keep it, please clean it up so it's better written.